pinot-core/src/main/java/org/apache/pinot/core/query/reduce/BrokerReduceService.java

```diff
@@ -527,4 +524,162 @@ private void setGroupByHavingResults(@Nonnull BrokerResponseNative brokerRespons
           "There should be minimum one aggregation function in the select list of a Group by query");
     }
   }
+
+  /**
+   * Following part are temporary code to handle the selection query with different DataSchema returned from different
+   * servers.
+   * TODO: Remove the code after all servers are migrated to the current version.
+   */
+
+  private static DataSchema getCanonicalDataSchema(Selection selection, DataSchema dataSchema) {
+    String[] columnNames = dataSchema.getColumnNames();
+    Map<String, Integer> columnToIndexMap = SelectionOperatorUtils.getColumnToIndexMap(columnNames);
+
+    Set<String> canonicalColumnSet = new HashSet<>();
```

Comment:
(nit): I think we know the exact size (as in the number of columns) so we should pass that as capacity to the constructor of these structures to prevent potential resize

Reply:
For HashSet, the initial capacity should not be configured the same as number of columns. The minimum capacity required should be (numColumns/loadFactor), but too small number can also increase the chance of hash collision.
Since the number of columns should be very few, the default size should be good enough IMO.

Reply:
Added the initial size for ArrayList